src/storage/compaction_test/src/server.rs

```diff
@@ -0,0 +1,282 @@
+    let join_handle = tokio::spawn(async move {
+        tokio::select! {
+            _ = &mut shutdown_recv => {
+                for (join_handle, shutdown_sender) in sub_tasks {
+                    if let Err(err) = shutdown_sender.send(()) {
+                        tracing::warn!("Failed to send shutdown: {:?}", err);
+                        continue;
+                    }
+                    if let Err(err) = join_handle.await {
+                        tracing::warn!("Failed to join shutdown: {:?}", err);
+                    }
+                }
+            },
+        }
+    });
+
+    let local_version_manager = hummock.local_version_manager();
+    // Replay version deltas from FIRST_VERSION_ID to the version before reset
+    let mut modified_compaction_groups = HashSet::<CompactionGroupId>::new();
+    let mut replay_count: u64 = 0;
+    let (start_version, end_version) = (FIRST_VERSION_ID + 1, version_before_reset.id + 1);
+    let mut replayed_epochs = vec![];
+    for id in start_version..end_version {
+        let (version_id, max_committed_epoch, compaction_groups) =
+            meta_client.replay_version_delta(id).await?;
+        tracing::info!(
+            "Replayed version delta version_id: {}, max_committed_epoch: {}, compaction_groups: {:?}",
+            version_id,
+            max_committed_epoch,
+            compaction_groups
+        );
+        replay_count += 1;
+        replayed_epochs.push(max_committed_epoch);
+        compaction_groups
+            .into_iter()
+            .map(|c| modified_compaction_groups.insert(c))
+            .count();
+
+        // We can custom more conditions for compaction triggering
+        // For now I just use a static way here
+        if replay_count % opts.compaction_trigger_frequency == 0
+            && !modified_compaction_groups.is_empty()
+        {
+            replayed_epochs.pop(); // pop the latest epoch
+            local_version_manager
+                .try_wait_epoch(HummockReadEpoch::Committed(max_committed_epoch))
+                .await?;
+
+            let mut epochs = vec![max_committed_epoch];
+            epochs.extend(
+                pin_old_snapshots(&meta_client, &mut replayed_epochs, 1)
+                    .await
+                    .into_iter(),
+            );
+
+            // Get kv pairs of snapshots
+            let expect_results = scan_epochs(&hummock, &epochs).await?;
+            tracing::info!(
+                "Trigger compaction for version {}, epoch {} compaction_groups: {:?}",
+                version_id,
+                max_committed_epoch,
+                modified_compaction_groups,
+            );
+            // When this function returns, the compaction is finished.
+            // The returned version deltas is order by id in ascending order
+            let (new_version_id, new_committed_epoch) = meta_client
+                .trigger_compaction_deterministic(
```

Comment:
It is better to trigger multiple rounds of compaction here. Otherwise, we may end up only testing L0 compaction with compaction concurrency = 1. We can address this in later PRs.

Reply:
I notice that there is a hashset in the `CompactionRequestChannel` to dedup request for a specific group. So I don't understand how we can have multiple compact tasks for a specific compaction group (compaction concurrency > 1)?

Reply:
Tasks is picked one by one, but there can be multiple inflight tasks. After successfully picking one task, scheduler will try to pick more.